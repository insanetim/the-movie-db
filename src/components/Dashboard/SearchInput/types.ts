export interface SearchInputProps {
  query: string
}

export interface SearchInputHookProps extends SearchInputProps {}

export interface SearchInputHook {
  currentValue: string
  handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  handleSearch: (value: string) => void
}
