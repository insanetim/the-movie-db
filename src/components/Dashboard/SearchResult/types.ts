import {
  ErrorMsg,
  HandlePagination,
  Query,
} from 'src/interfaces/global.interface'
import { IMoviesList } from 'src/interfaces/movie.interface'

export type SearchResultHookProps = SearchResultProps

export type SearchResultHookReturn = {
  error: ErrorMsg
  handlePagination: HandlePagination
  isLoading: boolean
  movies: IMoviesList | undefined
}

export type SearchResultProps = {
  query: Query
}
