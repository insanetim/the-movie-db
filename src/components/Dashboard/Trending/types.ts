import type { IMoviesList } from 'src/interfaces/movie.interface'

export interface TrendingHook {
  error: null | string
  handlePagination: (page: number) => void
  loading: boolean
  movies: IMoviesList | null
}
