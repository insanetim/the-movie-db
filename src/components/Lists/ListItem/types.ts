import { MouseEvent } from 'react'
import { IList } from 'src/interfaces/list.interface'

export type ListItemHookProps = {
  listId: IList['id']
  name: IList['name']
}

export type ListItemHookReturn = {
  handleClick: () => void
  handleListDelete: (event: MouseEvent<HTMLSpanElement>) => () => Promise<void>
}

export type ListItemProps = {
  description: IList['description']
  listId: IList['id']
  name: IList['name']
}
