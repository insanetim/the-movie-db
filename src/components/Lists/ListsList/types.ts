import type { IList } from 'src/interfaces/list.interface'

export interface ListsListProps {
  lists: IList[]
}
