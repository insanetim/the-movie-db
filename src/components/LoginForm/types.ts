import { UserData } from 'src/store/features/auth'

export type LoginFormHookReturn = {
  isDark: boolean
}

export type LoginFormProps = {
  isSubmitting: boolean
  onSubmit: (values: UserData) => void
}
