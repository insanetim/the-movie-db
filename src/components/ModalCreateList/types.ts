import { FormInstance, ModalProps } from 'antd'
import { IMovie } from 'src/interfaces/movie.interface'

export interface ModalCreateListProps extends ModalProps {
  movieId?: IMovie['id']
  onSuccess?: () => void
}

export interface ModalCreateListHookProps {
  form: FormInstance
  movieId?: IMovie['id']
  onSuccess?: () => void
}

export type ListData = { description: string; name: string }

export interface ModalCreateListHook {
  handleAfterClose: () => void
  handleOk: () => void
  handleSubmit: (listData: ListData) => void
}
