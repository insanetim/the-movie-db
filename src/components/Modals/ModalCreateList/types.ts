import { FormInstance, InputRef, ModalProps } from 'antd'
import { IMovie } from 'src/interfaces/movie.interface'

export type ListData = { description: string; name: string }

export type ModalCreateListHookProps = {
  form: FormInstance
  movieId?: IMovie['id']
  onSuccess?: () => void
}

export type ModalCreateListHookReturn = {
  handleAfterClose: () => void
  handleAfterOpenChange: (open: boolean) => void
  handleOk: () => void
  handleSubmit: (listData: ListData) => void
  inputRef: React.RefObject<InputRef>
}

export type ModalCreateListProps = ModalProps & {
  movieId?: IMovie['id']
  onSuccess?: () => void
}
