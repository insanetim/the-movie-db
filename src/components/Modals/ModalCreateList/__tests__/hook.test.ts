import { renderHook } from '@testing-library/react'
import { FormInstance, InputRef } from 'antd'
import { useRef } from 'react'
import { hideModal } from 'src/store/features/app'
import { ListData } from 'src/store/features/list'
import { useAppDispatch } from 'src/store/hooks'

import useContainer from '../hook'
import { ModalCreateListHookProps } from '../types'

jest.mock('src/store/hooks')
jest.mock('src/store/features/app', () => ({
  hideModal: jest.fn(),
}))

// Mock React useRef for testing inputRef
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: jest.fn(),
}))

const mockUseAppDispatch = useAppDispatch as jest.MockedFunction<
  typeof useAppDispatch
>
const mockHideModal = hideModal as jest.MockedFunction<typeof hideModal>
const mockUseRef = useRef as jest.MockedFunction<typeof useRef>

// Mock InputRef interface for testing
const mockInputRef: React.RefObject<InputRef> = {
  current: {
    blur: jest.fn(),
    focus: jest.fn(),
    input: null,
    select: jest.fn(),
    setSelectionRange: jest.fn(),
    value: '',
  } as unknown as InputRef,
}

// Mock FormInstance
const mockForm: FormInstance = {
  focusField: jest.fn(),
  getFieldError: jest.fn(),
  getFieldInstance: jest.fn(),
  getFieldsError: jest.fn(),
  getFieldsValue: jest.fn(),
  getFieldValue: jest.fn(),
  // Missing methods that are required by FormInstance interface
  getFieldWarning: jest.fn(),
  isFieldsTouched: jest.fn(),
  isFieldsValidating: jest.fn(),
  isFieldTouched: jest.fn(),
  isFieldValidating: jest.fn(),
  resetFields: jest.fn(),
  scrollToField: jest.fn(),
  setFields: jest.fn(),
  setFieldsValue: jest.fn(),
  setFieldValue: jest.fn(),
  submit: jest.fn(),
  validateFields: jest.fn(),
}

// Mock ListData
const mockListData: ListData = {
  description: 'Test Description',
  name: 'Test List',
}

describe('ModalCreateList useContainer hook', () => {
  let dispatch: jest.Mock
  let mockOnSubmit: jest.Mock

  beforeEach(() => {
    dispatch = jest.fn()
    mockUseAppDispatch.mockReturnValue(dispatch as never)

    mockUseRef.mockReturnValue(mockInputRef as never)

    mockHideModal.mockReturnValue({
      payload: undefined,
      type: 'app/hideModal',
    } as never)

    mockOnSubmit = jest.fn()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should return hook functions and refs', () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    expect(result.current.handleOk).toBeDefined()
    expect(result.current.handleSubmit).toBeDefined()
    expect(result.current.handleAfterOpenChange).toBeDefined()
    expect(result.current.inputRef).toBe(mockInputRef)
  })

  it('should call form.submit() when handleOk is called', () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    result.current.handleOk()

    expect(mockForm.submit).toHaveBeenCalledTimes(1)
  })

  it('should dispatch hideModal and call onSubmit when handleSubmit is called', async () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    await result.current.handleSubmit(mockListData)

    expect(dispatch).toHaveBeenCalledWith(mockHideModal())
    expect(mockOnSubmit).toHaveBeenCalledWith(mockListData)
  })

  it('should dispatch hideModal and call onSubmit even without onSubmit prop', async () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      // onSubmit is optional
    }

    const { result } = renderHook(() => useContainer(props))

    await result.current.handleSubmit(mockListData)

    expect(dispatch).toHaveBeenCalledWith(mockHideModal())
    expect(mockOnSubmit).not.toHaveBeenCalled()
  })

  it('should focus input when handleAfterOpenChange is called with open=true', () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    result.current.handleAfterOpenChange(true)

    expect(mockInputRef.current!.focus).toHaveBeenCalledTimes(1)
  })

  it('should not focus input when handleAfterOpenChange is called with open=false', () => {
    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    result.current.handleAfterOpenChange(false)

    expect(mockInputRef.current!.focus).not.toHaveBeenCalled()
  })

  it('should handle null input ref gracefully in handleAfterOpenChange', () => {
    mockUseRef.mockReturnValue({ current: null } as never)

    const props: ModalCreateListHookProps = {
      form: mockForm,
      onSubmit: mockOnSubmit,
    }

    const { result } = renderHook(() => useContainer(props))

    // Should not throw an error
    expect(() => result.current.handleAfterOpenChange(true)).not.toThrow()
  })
})
