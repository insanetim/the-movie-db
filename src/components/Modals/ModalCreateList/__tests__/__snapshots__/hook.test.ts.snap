// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalCreateList useContainer hook should match snapshot 1`] = `
{
  "handleAfterClose": [Function],
  "handleAfterOpenChange": [Function],
  "handleOk": [Function],
  "handleSubmit": [Function],
  "inputRef": {
    "current": null,
  },
}
`;
